### PR TITLE
Add integration marker, CI step + codecov flag

### PR DESCRIPTION
**Pull request by anthonyburdi**
Add an integration marker, CI step + codecov flag for tests that rely on a running GX Cloud container.

See request: https://github.com/great-expectations/cloud/pull/240#discussion_r1585517026

Note that this also marks `test_agent.py` as an integration test. We should look into those tests and aim to make them run as unit tests, e.g. by mocking calls to the backend.

Note also that I'm not sure about the codecov flag config here yet so keeping as draft.

### CODE REVIEW

File: tests/agent/test_agent.py [hunk 1]
```diff
@@ -39,6 +39,9 @@
     from tests.agent.conftest import DataContextConfigTD
 
 
+pytestmark = pytest.mark.integration
```
Comment (Shinnnyshinshin) [inline tests/agent/test_agent.py:42]: would these qualify as integration tests? I was thinking `unit`

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -40,6 +40,47 @@ jobs:
       - run: poetry run invoke ${{ matrix.task }} --check
 
   tests:
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.9", "3.10", "3.11"]
+    runs-on: ubuntu-latest
```
Comment (Shinnnyshinshin) [inline .github/workflows/ci.yaml:46]: 🚀

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -40,6 +40,47 @@ jobs:
...
+      - name: Run tests
+        run: poetry run pytest -m "not agentjobs and not integration" --cov-report xml
+
+      # upload coverage report to codecov
+      - name: Upload coverage reports to Codecov
+        uses: codecov/codecov-action@v4.1.0
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          slug: great-expectations/cloud
+          flags: ${{ matrix.python-version }},unit
```
Comment (anthonyburdi) [inline .github/workflows/ci.yaml]: I'm not sure I did the codecov flags correctly. Originally I had separate flags e.g. `flags: ${{ matrix.python-version }},unit` which shows as:
image
Then changed to more of a namespace-like approach: `flags: unit-${{ matrix.python-version }}` and `flags: integration-${{ matrix.python-version }}`
Feedback welcome!

File: tests/agent/test_agent.py (hunk 1 above)
Comment (anthonyburdi) [reply tests/agent/test_agent.py:42]: I thought so too, but they do require more than the env variables - it looks like they are trying to make a connection: https://github.com/great-expectations/cloud/actions/runs/8903149643/job/24450383609#step:7:726. I agree, after a brief look I think these should be unit tests but if so we probably need to change the tests themselves.

Comment (Kilo59) [reply tests/agent/test_agent.py:42]: Can you annotate the markers in `init_options` to explain what we are considering an integration test.

`integration: test that relies on a running GX Cloud instance`.

And/or maybe just make the marker name more explicit. `gx_cloud_integration`

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (Kilo59) [reply .github/workflows/ci.yaml]: I don't think we need to run these integration tests on every version of python.
The non-integration tests should catch any python version related problems.

I would just run the integration tests on a single version of python, and leave the python version out of the codecov flag. I would use whatever version we use for the container.

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -40,6 +40,47 @@ jobs:
       - run: poetry run invoke ${{ matrix.task }} --check
 
   tests:
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.9", "3.10", "3.11"]
```
Comment (Kilo59) [inline .github/workflows/ci.yaml:45]:
```suggestion
        python-version: ["3.11"]
```

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -40,6 +40,47 @@ jobs:
...
+      - name: Run tests
+        run: poetry run pytest -m "not agentjobs and not integration" --cov-report xml
+
+      # upload coverage report to codecov
+      - name: Upload coverage reports to Codecov
+        uses: codecov/codecov-action@v4.1.0
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          slug: great-expectations/cloud
+          flags: unit-${{ matrix.python-version }}
```
Comment (Kilo59) [inline .github/workflows/ci.yaml:81]:
```suggestion
          flags: ${{ matrix.python-version }}
```
The tests running are more than just unit-tests.
I would leave this out of the flag.

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -40,6 +40,47 @@ jobs:
...
+
+  integration-tests:
     strategy:
       matrix:
         python-version: ["3.8", "3.9", "3.10", "3.11"]
```
Comment (Kilo59) [inline .github/workflows/ci.yaml:86]:
```suggestion
        python-version: ["3.10"]
```

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -111,15 +152,15 @@ jobs:
...
+        run: poetry run pytest -m integration -W default::great_expectations.datasource.fluent.GxInvalidDatasourceWarning --cov-report xml
 
       # upload coverage report to codecov
       - name: Upload coverage reports to Codecov
         uses: codecov/codecov-action@v4.1.0
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
           slug: great-expectations/cloud
-          flags: ${{ matrix.python-version }}
+          flags: integration-${{ matrix.python-version }}
```
Comment (Kilo59) [inline .github/workflows/ci.yaml:163]:
```suggestion
          flags: integration
```

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -111,15 +152,15 @@ jobs:
 
       - name: Run tests
 
-        run: poetry run pytest -m "not agentjobs" -W default::great_expectations.datasource.fluent.GxInvalidDatasourceWarning --cov-report xml
+        run: poetry run pytest -m integration -W default::great_expectations.datasource.fluent.GxInvalidDatasourceWarning --cov-report xml
```
Comment (Kilo59) [inline .github/workflows/ci.yaml:155]:
```suggestion
        run: poetry run pytest -vv -m integration -W default::great_expectations.datasource.fluent.GxInvalidDatasourceWarning --cov-report xml
```
We shouldn't have that many integration tests so I think using the `verbose` flag can be helpful especially when these tests get slower.

**Review by Kilo59 — APPROVED** [review]

**Comment by Kilo59** [comment]
Thanks for this!

File: tests/agent/test_agent.py (hunk 1 above)
Comment (anthonyburdi) [reply tests/agent/test_agent.py:42]: Do you happen to know what the `agentjobs` marker is for? It looks like a single test `test_job_processing` which looks like it runs a checkpoint job during the docker image build pipeline. I'm guessing a integration/smoketest?

**Pull request merged**